charms/istio-pilot/src/charm.py

```diff
@@ -828,7 +894,101 @@ def _istiod_svc(self):
+                "ssl-key": base64.b64encode(
+                    ssl_secret.get_content()["ssl-key"].encode("ascii")
+                ).decode("utf-8"),
+            }
+        except SecretNotFoundError:
+            return {
+                "ssl-crt": base64.b64encode(self._cert_handler.cert.encode("ascii")).decode(
+                    "utf-8"
+                ),
+                "ssl-key": base64.b64encode(self._cert_handler.key.encode("ascii")).decode(
+                    "utf-8"
+                ),
+            }
+
+    # ---- Start of the block
+    # ---- WARNING: this feature is not recommended, but is supported in 1.17 and 1.18.
+    # ---- For details please refer to canonical/istio-operators#380.
+    # ---- FIXME: Remove this block after releasing 1.18.
+    def _use_https(self) -> bool:
+        """Return True if only one of the TLS configurations are enabled, False if none are.
+
+        Raises:
+            ErrorWithStatus: if both configurations are enabled.
+        """
+        if self._use_https_with_tls_provider() and self._use_https_with_tls_secret():
+            self.log.error(
+                "Only one TLS configuration is supported at a time."
+                "Either remove the TLS certificate provider relation,"
+                "or the TLS manual configuration with the remove-tls-secret action."
+            )
+            raise ErrorWithStatus(
+                "Only one TLS configuration is supported at a time. See logs for details.",
+                BlockedStatus,
+            )
+        return self._use_https_with_tls_provider() or self._use_https_with_tls_secret()
+
+    def _use_https_with_tls_secret(self) -> bool:
+        """Return True if both SSL key and crt are set with save-tls-secret, False otherwise.
+
+        Raises:
+            ErrorWithStatus: if one of the values is missing.
+        """
+
+        # Ensure the secret that holds the values exist otherwise we can assume
+        # that users don't want to configure TLS or they want to do it via a TLS provider.
+        try:
+            secret = self.model.get_secret(label="istio-tls-secret")
+        except SecretNotFoundError:
+            return False
+
+        ssl_key = secret.get_content()["ssl-key"]
+        ssl_crt = secret.get_content()["ssl-crt"]
+
+        # Fail if ssl is only partly configured as this is probably a mistake
+        # In reality this part of the code should never be executed as the action
+        # sets these values as required, but leaving the check in case somewhere in
+        # the code these values are messed up.
+        if _xor(ssl_key, ssl_crt):
+            missing = "ssl-key"
+            if not secret.get_content()["ssl-crt"]:
+                missing = "ssl-crt"
+            raise ErrorWithStatus(f"Missing {missing}, cannot configure TLS.", BlockedStatus)
+
+        # If both the SSL key and cert are provided, we configure TLS
+        return True if ssl_key and ssl_crt else False
```

Comment:
Is there a reason why this:
* returns `False` if both `cert` and `key` are unset
* goes to blocked if one of `cert` and `key` are unset

?

Having the secret set and `cert`/`key` unset sounds like its still a problem.  In this case, TLS is requested but not configured.  Returning False here means the gateway will be set up without auth, which probably isn't what is desired

Reply:
Updated, no reason to return False here, we just want to return True iff both are set.

Reply:
Yeah that's fine with me.  Especially where this is a short-lived feature anyway so not likely someone changes the action by mistake in future

Reply:
It's updated, please let me know what you think (: